src/imitation/algorithms/adversarial/common.py

```diff
@@ -427,6 +430,45 @@ def _torchify_array(self, ndarray: Optional[np.ndarray]) -> Optional[th.Tensor]:
         if ndarray is not None:
             return th.as_tensor(ndarray, device=self.reward_train.device)
 
+    def _get_log_policy_act_prob(
+        self,
+        obs_th: th.Tensor,
+        acts_th: th.Tensor,
+    ) -> th.Tensor:
+        """Evaluates the given actions on the given observations.
+
+        Args:
+            obs: A batch of observations.
+            actions: A batch of actions.
+
+        Returns:
+            A batch of log policy action probabilities.
+        """
+        log_policy_act_prob = None
+        if isinstance(self.gen_algo.policy, policies.ActorCriticPolicy):
+            # policies.ActorCriticPolicy has a concrete implementation of
+            # evaluate_actions to generate log_policy_act_prob given obs and actions.
+            _, log_policy_act_prob_th, _ = self.gen_algo.policy.evaluate_actions(
+                obs_th,
+                acts_th,
+            )
+        elif isinstance(self.gen_algo.policy, sac_policies.SACPolicy):
+            gen_algo_actor = self.gen_algo.policy.actor
+            # generate log_policy_act_prob from SAC actor.
+            mean_actions, log_std, _ = gen_algo_actor.get_action_dist_params(obs_th)
+            distribution = gen_algo_actor.action_dist.proba_distribution(
+                mean_actions,
+                log_std,
+            )
+            log_policy_act_prob_th = distribution.log_prob(acts_th)
+            # getting log probability of actions from a continuous distribution
```

Comment:
Under what situation have you run into this being a problem? PPO also has a continuous distribution and we've tested on continuous environments and AFAIK not seen any NaNs. `evaluate_actions` in https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/common/policies.py#L632 does not seem to do any clipping.

Reply:
Oh, you're correct. I misidentified the problem.

- `nan` doesn't come from producing `log_prob` for actions far away from the distribution mean.
- The problem comes from the squashing function used by SAC (mentioned in [SpinningUp](https://spinningup.openai.com/en/latest/algorithms/sac.html#id1), [original paper Appendix C](https://arxiv.org/pdf/1801.01290.pdf), and [SB3 implementation](https://github.com/DLR-RM/stable-baselines3/blob/d68f0a2411766beb6da58ee0e989d1a6a72869bc/stable_baselines3/common/distributions.py#L195)) Specifically `SquashedDiagGaussianDistribution` squashes the sampled actions between (-1, 1) from a DiagGaussianDistribution by passing through a tanh function, hence it's expected that the actions are within (-1, 1).
- Then when [calculating log_prob](https://github.com/DLR-RM/stable-baselines3/blob/d68f0a2411766beb6da58ee0e989d1a6a72869bc/stable_baselines3/common/distributions.py#L225), the potentially unbounded actions could pass through `th.log(1-action**2+epsilon)`, which results in `nan` when `abs(actions) > 1`. This does happen for [pendulum](https://github.com/openai/gym/blob/053ee809324861dc1bec19feee97b47ba1557a94/gym/envs/classic_control/pendulum.py#L38) with its `action_space` being `Box(-2, 2)`.
- To be more specific, `acts_th` directly came from [`self._gen_replay_buffer`](https://github.com/HumanCompatibleAI/imitation/blob/2db81fec1490ea4624881fc44b9b566626f0658a/src/imitation/algorithms/adversarial/common.py#L459), which collects `gen_samples` with unscaled actions.

Proposed solutions:
- SB3.SAC does scale the [randomly sampled actions](https://github.com/DLR-RM/stable-baselines3/blob/d68f0a2411766beb6da58ee0e989d1a6a72869bc/stable_baselines3/common/off_policy_algorithm.py#L407) using `self.policy.scale_action`. I believe this would also work for our case.